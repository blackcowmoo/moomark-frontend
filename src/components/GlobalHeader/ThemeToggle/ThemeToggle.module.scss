
.checkbox {
  opacity: 0;
  position: absolute;
}

.container{
  width: 3rem;
}

.label {
  background-color: var(--bg-primary);
  border-radius: 50px;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 5px;
  position: relative;
  height: 26px;
  width: 50px;
  transform: scale(1.5);
  .ball {
    background-color: var(--secondary);
    border-radius: 50%;
    position: absolute;
    top: 2px;
    left: 2px;
    height: 22px;
    width: 22px;
    transform: translateX(0px);
    transition: transform 0.2s linear;
  }
  .moon {
    display: flex;
    width: 1rem;
    height: 1rem;
  }

  .sun {
    display: flex;
    width: 1rem;
    height: 1rem;
  }
}
