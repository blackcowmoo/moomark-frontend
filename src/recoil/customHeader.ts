import { atom, selector } from 'recoil';

export const defaultHeaderOption = { 'x-moom-route': '' };

export const customHeaderAtom = atom<Object>({
  key: 'customheader',
  default: defaultHeaderOption,
});

export const customHeaderState = selector({
  key: 'customHeader',
  get: ({ get }) => {
    const customHeader = get(customHeaderAtom);
    return customHeader;
  },
});
