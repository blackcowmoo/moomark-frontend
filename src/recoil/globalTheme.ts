import { atom } from 'recoil';

export type Theme = 'light' | 'dark' | null;

export const globalThemeState = atom<Theme>({
  key: 'globalThemeState',
  default: null,
});
