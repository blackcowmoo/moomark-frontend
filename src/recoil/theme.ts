import { atom } from 'recoil';

export type Theme = 'light' | 'dark' | null;

export const themeState = atom<Theme>({
  key: 'globalThemeState',
  default: null,
});
