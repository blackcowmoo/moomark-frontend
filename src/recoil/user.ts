import { atom } from 'recoil';

export interface IUser {
  nickname: string;
  email: string;
  picture: null | string;
  role: string | null;
}

export const initialUserState: IUser = { nickname: '', email: '', picture: null, role: null };

export const userRecoilState = atom<IUser>({
  key: 'currentUser',
  default: initialUserState,
});
