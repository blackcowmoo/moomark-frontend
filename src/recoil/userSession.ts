import { atom, selector } from 'recoil';

interface IUserSession {
  id: null | string;
  userName: null | string;
}

export const userSessionAtom = atom<IUserSession>({
  key: 'userSession',
  default: { id: null, userName: null },
});

export const loginUserState = selector({
  key: 'loginUserState',
  get: ({ get }) => {
    const userSession = get(userSessionAtom);
    const { userName } = userSession;

    return { userName };
  },
});
