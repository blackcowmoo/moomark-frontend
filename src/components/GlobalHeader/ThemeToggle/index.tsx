import { useEffect } from 'react';
import { useRecoilState } from 'recoil';
import { globalThemeState, Theme } from 'recoil/globalTheme';

import SunIcon from './Sun.svg';
import MoonIcon from './Moon.svg';

import styles from './ThemeToggle.module.scss';


const index = () => {
  const [theme, setTheme] = useRecoilState(globalThemeState);

  const isTheme = (value: any): value is Theme => {
    return typeof value === 'string' && (value === 'light' || value === 'dark') && typeof value !== 'undefined';
  };

  useEffect(() => {
    const htmlTheme = document.body.dataset.theme;
    const theme = isTheme(htmlTheme) ? htmlTheme : 'light';
    setTheme(theme);
  }, []);

  useEffect(() => {
    if (theme) {
      document.body.dataset.theme = theme;
      localStorage.setItem('theme', theme);
    }
  }, [theme]);

  const toggleTheme = () => {
    setTheme(theme === 'light' ? 'dark' : 'light');
  };

  return (
    <div className={styles.container}>
      <input type='checkbox' className={styles.checkbox} onChange={toggleTheme} checked={theme === 'dark'} id='chk' />
      {theme && (
        <label className={styles.label} htmlFor='chk'>
          <div className={styles.sun}>
            <SunIcon />
          </div>
          <div className={styles.moon}>
            <MoonIcon />
          </div>
          <div className={styles.ball} style={{ transform: `translateX(${theme === 'dark' ? '0px' : '24px'})` }}></div>
        </label>
      )}
    </div>
  );
};

export default index;
