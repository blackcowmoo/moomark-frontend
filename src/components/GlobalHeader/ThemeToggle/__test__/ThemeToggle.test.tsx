import { mount } from 'enzyme';
import { RecoilRoot } from 'recoil';
import ThemeToggle from '../index';

describe('<ThemeToggle/>', () => {
  it('matches snapshot', () => {
    const wrapper = mount(
      <RecoilRoot>
        <ThemeToggle />
      </RecoilRoot>,
    );
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
