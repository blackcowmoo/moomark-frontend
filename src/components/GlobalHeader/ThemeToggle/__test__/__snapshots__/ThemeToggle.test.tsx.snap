// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThemeToggle/> matches snapshot 1`] = `
"<RecoilRoot>
  <RecoilRoot_INTERNAL>
    <Batcher setNotifyBatcherOfChange={[Function (anonymous)]} />
    <index>
      <div className=\\"container\\">
        <input type=\\"checkbox\\" className=\\"checkbox\\" onChange={[Function: toggleTheme]} checked={false} id=\\"chk\\" />
        <label className=\\"label\\" htmlFor=\\"chk\\">
          <div className=\\"sun\\">
            <test-file-stub />
          </div>
          <div className=\\"moon\\">
            <test-file-stub />
          </div>
          <div className=\\"ball\\" style={{...}} />
        </label>
      </div>
    </index>
  </RecoilRoot_INTERNAL>
</RecoilRoot>"
`;
